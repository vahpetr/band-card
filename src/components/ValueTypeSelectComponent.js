// @ts-check

import React from "react";
import { ValueType } from "../types";

export class ValueTypeSelectComponent extends React.Component {
  availableValueTypes = [ValueType.YIELD, ValueType.SPREAD, ValueType.PRICE];

  render() {
    const { filter } = this.props;
    return (
      <select defaultValue={filter.valueType} onChange={this.onValueTypeChange}>
        {this.buildOptions()}
      </select>
    );
  }

  onValueTypeChange = ev => {
    this.props.onFilterChange({
      valueType: ev.target.value
    });
  };

  buildOptions() {
    return this.availableValueTypes.map(availableValueType => (
      <option value={availableValueType} key={availableValueType}>
        {availableValueType}
      </option>
    ));
  }
}
