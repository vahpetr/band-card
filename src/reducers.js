// @ts-check

import { combineReducers } from "redux";
import {
  BOND_DATA_FETCHING,
  BOND_DATA_SUCCEED,
  BOND_DATA_FAILED
} from "./actions";
import { ValueType } from "./types";

/**
 * App initial state
 */
export const initialState = Object.freeze({
  bond: {
    filter: {
      isin: "RU000A0JU4L3",
      valueType: ValueType.PRICE
    },
    data: [],
    error: null
  }
});

/**
 * Card reducer
 */
export function cardReducer(state = initialState, action) {
  switch (action.type) {
    case BOND_DATA_FETCHING: {
      const newState = { ...state };
      const newFilter = { ...newState.bond.filter, ...action.payload };
      newState.bond.filter = newFilter;
      return newState;
    }
    case BOND_DATA_SUCCEED: {
      const newState = { ...state };
      newState.bond.data = action.payload.data;
      newState.bond.error = null;
      return newState;
    }
    case BOND_DATA_FAILED: {
      const newState = { ...state };
      newState.bond.data = [];
      newState.bond.error = action.payload;
      return newState;
    }
    default: {
      return state;
    }
  }
}

/**
 * App reducer. All combinded reducers
 */
export const appReducers = combineReducers({
  cardReducer
});
