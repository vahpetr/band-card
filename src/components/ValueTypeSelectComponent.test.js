import React from "react";
import { configure, mount } from "enzyme";
import sinon from 'sinon';
import { shallowToJson } from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import { initialState } from "../reducers";
import { ValueTypeSelectComponent } from "./ValueTypeSelectComponent";

configure({ adapter: new Adapter() });

describe("<ValueTypeSelectComponent />", () => {
  it("renders without crashing", () => {
    const { filter } = initialState.bond;
    const onBondDataFilterChange = sinon.spy();
    const wrapper = mount(
      <ValueTypeSelectComponent
        filter={filter}
        onFilterChange={onBondDataFilterChange}
      />
    );

    expect(wrapper.props().filter.isin).toBe(filter.isin);
    expect(wrapper.props().filter.valueType).toBe(filter.valueType);
    expect(wrapper.props().onFilterChange).toBe(onBondDataFilterChange);
  });

  it("snapshot", () => {
    const { filter } = initialState.bond;
    const onBondDataFilterChange = sinon.spy();
    const wrapper = mount(
      <ValueTypeSelectComponent
        filter={filter}
        onFilterChange={onBondDataFilterChange}
      />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('simulates click events', () => {
    const { filter } = initialState.bond;
    const onBondDataFilterChange = sinon.spy();
    const wrapper = mount(
      <ValueTypeSelectComponent
        filter={filter}
        onFilterChange={onBondDataFilterChange}
      />
    );
    wrapper.simulate('change');
    expect(onBondDataFilterChange.called).toBeTruthy();
  });
});
