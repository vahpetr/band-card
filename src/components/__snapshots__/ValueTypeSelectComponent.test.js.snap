// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ValueTypeSelectComponent /> snapshot 1`] = `
<ValueTypeSelectComponent
  filter={
    Object {
      "isin": "RU000A0JU4L3",
      "valueType": "Price",
    }
  }
  onFilterChange={[Function]}
>
  <select
    defaultValue="Price"
    onChange={[Function]}
  >
    <option
      key="Yield"
      value="Yield"
    >
      Yield
    </option>
    <option
      key="Spread"
      value="Spread"
    >
      Spread
    </option>
    <option
      key="Price"
      value="Price"
    >
      Price
    </option>
  </select>
</ValueTypeSelectComponent>
`;
